src/coreclr/pal/inc/pal.h

```diff
@@ -3639,24 +3639,42 @@ Define_InterlockMethod(
     CHAR,
     InterlockedExchange8(IN OUT CHAR volatile *Target, CHAR Value),
     InterlockedExchange8(Target, Value),
+#ifdef __clang__
+    return __sync_swap(pDst, iValue); 
+#elif ATOMIC_CHAR_LOCK_FREE == 2
```

Comment:
We cannot depend on defines like this in CoreCLR PAL headers. CoreCLR PAL headers cannot depend on system headers where they are defined.

Reply:
Would moving the checks for those to the FCall definitions make them okay?

Reply:
No. Nothing in CoreCLR outside the PAL implementation can depend on system headers.

One way to fix this is to use `__sync_swap` clang-specific intrinsic for clang, and implement it using `__sync_val_compare_and_swap` for non-clang. Like:

```
#ifdef __clang__

Define_InterlockMethod(
    LONGLONG,
    InterlockedExchange64(IN OUT LONGLONG volatile *Target, IN LONGLONG Value),
    InterlockedExchange64(Target, Value),
    __sync_swap(pDst, iValue)
)

#else

inline LONGLONG InterlockedExchange64(LONGLONG volatile * Target, LONGLONG Value)
{
    LONGLONG Old;
    do {
        Old = *Target;
    } while (__sync_val_compare_and_swap(Target, Old, Value) != Old);

    PAL_InterlockedOperationBarrier();

    return Old;
}

#endif
```

Reply:
`__sync_val_compare_and_swap` isn't guaranteed to be lock free in GCC like we've discussed earlier (GCC docs it's implemented with `__atomic` in fact) so that wouldn't help here.
Maybe the test for this could just be done in the CMake files completely instead?

Reply:
I understand that it is not guaranteed, but it is reasonable to assume that the implementation is lock-free on platforms that we care about.

How would you do that in CMake files?

clang is fine. I would like to avoid some complicated solution that is just for gcc. As I have said earlier, we do not have any testing for gcc-based builds, so there is no way to validate that it actually works.

Reply:
> How would you do that in CMake files?

Pushed a commit, a part of it is commented out cause I want to verify on the CI.

> As I have said earlier, we do not have any testing for gcc-based builds, so there is no way to validate that it actually works.

I've seen that some Linux distros ship GCC built dotnet in their repositories so I'd prefer to guarantee its working for them.

Reply:
![image](https://github.com/dotnet/runtime/assets/35800402/93ac6a14-32ae-42aa-a5a3-f90d64ee5bb0)
Seems working I think.